sdk/scheduler/src/main/java/com/mesosphere/sdk/curator/CuratorPersister.java

```diff
@@ -198,6 +198,102 @@ public static Builder newBuilder(String serviceName, String zookeeperHostPort) {
+            try {
+                result.put(unprefixedPath, client.getData().forPath(path));
+            } catch (KeeperException.NoNodeException e) {
+                result.put(unprefixedPath, null);
+            } catch (Exception e) {
+                throw new PersisterException(Reason.STORAGE_ERROR,
+                        String.format("Unable to retrieve data from %s", path), e);
+            }
+        }
+        return result;
+    }
+
+    @Override
+    public void setMany(Map<String, byte[]> unprefixedPathBytesMap) throws PersisterException {
+        if (unprefixedPathBytesMap.isEmpty()) {
+            return;
+        }
+        // Convert map to translated prefixed paths. Manually construct TreeMap for consistent ordering (for tests):
+        Map<String, byte[]> pathBytesMap = new TreeMap<>();
+        for (Map.Entry<String, byte[]> entry : unprefixedPathBytesMap.entrySet()) {
+            pathBytesMap.put(withFrameworkPrefix(entry.getKey()), entry.getValue());
+        }
+        LOGGER.debug("Updating {} entries: {}", pathBytesMap.size(), pathBytesMap.keySet());
+        runTransactionWithRetries(new SetTransactionFactory(pathBytesMap));
+    }
+
+    @Override
+    public void recursiveCopy(String unprefixedSrc, String unprefixedDest) throws PersisterException {
+        if (new HashSet<>(Arrays.asList(
+                serviceRootPath,
+                CuratorLocker.LOCK_PATH_NAME,
+                unprefixedSrc,
+                unprefixedDest)).size() != 4) {
+            throw new IllegalArgumentException(String.format("Cannot copy from %s to %s",
+                    unprefixedSrc, unprefixedDest));
+        }
+
+        try {
+            LOGGER.debug("Copying {} (and any children) to {}", unprefixedSrc, unprefixedDest);
+            if (client.checkExists().forPath(withFrameworkPrefix(unprefixedSrc)) == null) {
+                throw new PersisterException(Reason.NOT_FOUND,
+                        String.format("Source node does not exist: %s", unprefixedSrc));
+            }
+            if (client.checkExists().forPath(withFrameworkPrefix(unprefixedDest)) != null) {
+                throw new PersisterException(Reason.LOGIC_ERROR,
+                        String.format("Destination exists: %s", unprefixedDest));
+            }
+
+            LinkedList<String> toBeWalked = new LinkedList<>(Collections.singleton(unprefixedSrc));
+            Map<String, byte[]> toBeAdded = new HashMap<>();
+            while (!toBeWalked.isEmpty()) {
+                String nextNode = toBeWalked.poll();
+                // This assertion should never fail acc. to the logic around it.
+                assert (nextNode.startsWith(unprefixedSrc)) : String.format("Failed to match prefix." +
+                        " Src [%s] Dest [%s] Child [%s]", unprefixedSrc, unprefixedDest, nextNode);
+                toBeAdded.put(withFrameworkPrefix(nextNode.replace(unprefixedSrc, unprefixedDest)), get(nextNode));
```

Comment:
I agree. But i also feel that `curNode.replace(unprefixedSrc, unprefixedDest)` is "ugly" by itself but a good way to make sure this "ugly" code works is this should be inside this loop where we _definitely_ know it would not fail because of the logic around it. anyway, i will just keep this as is and address in a future cleanup when we end up sharing some code between `recursiveCopy` and `recursiveDelete`.